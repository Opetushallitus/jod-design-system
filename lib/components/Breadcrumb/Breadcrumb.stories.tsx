import type { StoryObj } from '@storybook/react-vite';
import type { TitledMeta } from '../../utils';
import { Breadcrumb } from './Breadcrumb';

const meta = {
  title: 'Navigation/Breadcrumb',
  component: Breadcrumb,
  tags: ['autodocs'],
} satisfies TitledMeta<typeof Breadcrumb>;

export default meta;

type Story = StoryObj<typeof meta>;

const parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/design/a3heEFNpoI9MRcMFtbYVfO/Osaamispolku---Design-system?node-id=356-2883',
  },
};

export const Default: Story = {
  parameters: {
    ...parameters,
    docs: {
      description: {
        story: 'Breadcrumb component for navigation.',
      },
    },
  },
  argTypes: {
    serviceVariant: {
      control: { type: 'radio' },
      options: ['yksilo', 'ohjaaja', 'tietopalvelu', 'palveluportaali'],
    },
  },
  args: {
    items: [
      { label: 'Etusivu', to: '/' },
      { label: 'Osaamisprofiili', to: '/osaamispolkuni' },
      { label: 'TyÃ¶paikkani' },
    ],
    LinkComponent: ({ to, children }) => <a href={to}>{children}</a>,
    ariaLabel: 'Breadcrumb',
    serviceVariant: 'yksilo',
  },
};
