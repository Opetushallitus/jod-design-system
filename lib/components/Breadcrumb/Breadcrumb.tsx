import React from 'react';
import { cx } from '../../cva';
import { type ServiceVariant } from '../../utils';

export interface BreadcrumbItem {
  id?: string;
  label: string;
  to?: string;
}

export interface BreadcrumbProps {
  items: BreadcrumbItem[];
  LinkComponent: React.ComponentType<{ to: string; children: React.ReactNode; className?: string }>;
  ariaLabel?: string;
  serviceVariant: ServiceVariant;
  dataTestId?: string;
}

export const Breadcrumb = ({ items, ariaLabel, LinkComponent, serviceVariant, dataTestId }: BreadcrumbProps) => {
  const textColorClass = cx({
    'ds:text-secondary-1-dark': serviceVariant === 'yksilo',
    'ds:text-secondary-2-dark': serviceVariant === 'ohjaaja',
    'ds:text-secondary-3-dark': serviceVariant === 'palveluportaali',
    'ds:text-secondary-4-dark': serviceVariant === 'tietopalvelu',
  });
  return (
    <nav
      aria-label={ariaLabel}
      className={`ds:text-accent ds:text-[12px] ds:font-bold ds:col-span-3 ds:mb-3 ds:print:hidden ${textColorClass}`}
      data-testid={dataTestId}
    >
      <ol className="ds:flex ds:flex-wrap ds:gap-y-2">
        {items.map((item, idx) => {
          const isLast = idx === items.length - 1;

          return (
            <li
              key={item.id ?? item.to ?? item.label}
              className="ds:not-first:before:content-['/'] ds:before:mx-2 ds:before:text-secondary-5"
            >
              {item.to && !isLast ? (
                <span className="ds:underline">
                  <LinkComponent to={item.to}>{item.label}</LinkComponent>
                </span>
              ) : (
                <span className="ds:text-black" aria-current="location">
                  {item.label}
                </span>
              )}
            </li>
          );
        })}
      </ol>
    </nav>
  );
};
