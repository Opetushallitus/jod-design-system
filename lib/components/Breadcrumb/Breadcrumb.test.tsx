import { render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { Breadcrumb } from './Breadcrumb';

const items = [
  { label: 'Etusivu', to: '/' },
  { label: 'Osaamisprofiili', to: '/osaamispolkuni' },
  { label: 'Työpaikkani' },
];

const LinkComponent = ({ to, children }: { to?: string; children: React.ReactNode }) => <a href={to}>{children}</a>;

describe('Breadcrumb', () => {
  it('renders all breadcrumb items', () => {
    const { container } = render(
      <Breadcrumb items={items} LinkComponent={LinkComponent} ariaLabel="Breadcrumb" serviceVariant="yksilo" />,
    );
    expect(screen.getByText('Etusivu')).toBeInTheDocument();
    expect(screen.getByText('Osaamisprofiili')).toBeInTheDocument();
    expect(screen.getByText('Työpaikkani')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders links for items with "to" prop', () => {
    render(<Breadcrumb items={items} LinkComponent={LinkComponent} ariaLabel="Breadcrumb" serviceVariant="yksilo" />);
    const link = screen.getByText('Etusivu').closest('a');
    expect(link).toHaveAttribute('href', '/');
  });

  it('does not render link for last item if "to" is missing', () => {
    render(<Breadcrumb items={items} LinkComponent={LinkComponent} ariaLabel="Breadcrumb" serviceVariant="yksilo" />);
    const last = screen.getByText('Työpaikkani');
    expect(last.closest('a')).toBeNull();
  });

  it('sets aria-label for navigation', () => {
    render(<Breadcrumb items={items} LinkComponent={LinkComponent} ariaLabel="Custom label" serviceVariant="yksilo" />);
    const nav = screen.getByLabelText('Custom label');
    expect(nav).toBeInTheDocument();
  });

  it('emits data-testid when dataTestId is provided', () => {
    render(
      <Breadcrumb
        items={items}
        LinkComponent={LinkComponent}
        ariaLabel="Crumbs"
        serviceVariant="yksilo"
        dataTestId="crumbs"
      />,
    );
    expect(screen.getByTestId('crumbs')).toBeInTheDocument();
  });
});
