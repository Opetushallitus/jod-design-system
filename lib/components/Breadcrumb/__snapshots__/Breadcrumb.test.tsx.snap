// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Breadcrumb > renders all breadcrumb items 1`] = `
<nav
  aria-label="Breadcrumb"
  class="ds:text-primary-gray ds:font-poppins ds:text-card-label ds:col-span-3 ds:h-5 ds:print:hidden ds:content-center"
>
  <ol
    class="ds:flex ds:flex-wrap ds:gap-y-2"
  >
    <li
      class="ds:flex ds:items-center"
    >
      <span
        class="ds:hover:underline"
      >
        <a
          href="/"
        >
          Etusivu
        </a>
      </span>
    </li>
    <li
      class="ds:flex ds:items-center"
    >
      <span
        aria-hidden="true"
        class="ds:mx-2 ds:text-secondary-5"
      >
        /
      </span>
      <span
        class="ds:hover:underline"
      >
        <a
          href="/osaamispolkuni"
        >
          Osaamisprofiili
        </a>
      </span>
    </li>
    <li
      class="ds:flex ds:items-center"
    >
      <span
        aria-hidden="true"
        class="ds:mx-2 ds:text-secondary-5"
      >
        /
      </span>
      <span
        aria-current="page"
        class="ds:text-secondary-1-dark"
      >
        Ty√∂paikkani
      </span>
    </li>
  </ol>
</nav>
`;
